src/video.cpp

```diff
@@ -191,8 +194,98 @@ void CVideo::update_framebuffer()
 		return;
 	}
 
-	surface fb = SDL_GetWindowSurface(*window);
-	frameBuffer = fb;
+	// Find max valid pixel scale at current window size.
+	point wsize(window->get_size());
+	int max_scale = std::min(
+		wsize.x / preferences::min_window_width,
+		wsize.y / preferences::min_window_height);
+	max_scale = std::min(max_scale, preferences::max_pixel_scale);
+
+	// Determine best pixel scale according to preference and window size
+	int scale = 1;
+	if (preferences::auto_pixel_scale()) {
+		// Try to match the default size (1280x720) but do not reduce below
+		int def_scale = std::min(
+			wsize.x / preferences::def_window_width,
+			wsize.y / preferences::def_window_height);
+		scale = std::min(max_scale, def_scale);
+		// Otherwise reduce to keep below the max window size (1920x1080).
+		int min_scale = std::min(
+			wsize.x / (preferences::max_window_width+1) + 1,
+			wsize.y / (preferences::max_window_height+1) + 1);
+		scale = std::max(scale, min_scale);
+	} else {
+		scale = std::min(max_scale, preferences::pixel_scale());
+	}
+
+	// Update logical size if it doesn't match the current resolution and scale.
+	point lsize(window->get_logical_size());
+	point osize(window->get_output_size());
+	if (lsize.x != wsize.x / scale || lsize.y != wsize.y / scale) {
+		if (!preferences::auto_pixel_scale() && scale < preferences::pixel_scale()) {
+			LOG_DP << "reducing pixel scale from desired "
+				<< preferences::pixel_scale() << " to maximum allowable "
+				<< scale << std::endl;
+		}
+		LOG_DP << "pixel scale: " << scale << std::endl;
+		LOG_DP << "overriding logical size" << std::endl;
+		LOG_DP << "  old lsize: " << lsize << std::endl;
+		LOG_DP << "  old wsize: " << wsize << std::endl;
+		LOG_DP << "  old osize: " << osize << std::endl;
+		window->set_logical_size(osize.x / scale, osize.y / scale);
+		lsize = window->get_logical_size();
+		wsize = window->get_size();
+		osize = window->get_output_size();
+		LOG_DP << "  new lsize: " << lsize << std::endl;
+		LOG_DP << "  new wsize: " << wsize << std::endl;
+		LOG_DP << "  new osize: " << osize << std::endl;
+	}
+
+	// Update the drawing surface if required.
+	if (!drawingSurface
+		|| drawingSurface->w != wsize.x / scale
+		|| drawingSurface->h != wsize.y / scale)
+	{
+		uint32_t format = window->pixel_format();
+		int bpp = SDL_BITSPERPIXEL(format);
+
+		// Free the old drawing surface.
+		if (drawingSurface) {
+			// Except it can still be in use... what should we do about that?
+			LOG_DP << "leaking old drawing surface" << std::endl;
+			//LOG_DP << "freeing old drawing surface" << std::endl;
+			//SDL_FreeSurface(drawingSurface);
```

Comment:
Is that a TODO before merging the PR? IDK whom to summon from the wesnoth team here :thinking:

Reply:
Not really. It's only ever called when changing window size so it shouldn't happen frequently enough to be a problem. But i don't like it so i left the comment there.

I tried cleaning up properly but something elsewhere was still accessing the old surface and it looked like hell to debug (and would probably open a whole can of multithreading worms).

Reply:
Oh, derp. Turns out it does already have a wrapper which is freeing it in the destructor. Just tested and it seems to be doing the right thing. I'll just remove this block.